cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,103 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CloudPrivateIPConfig allows the user to assign private IP addresses
+// to the primary NIC associated with cloud VMs. This is done by
+// specifying a list Kubernetes nodes and IPs requested for those nodes.
+// This CRD is intended on being used by the network plugin which manages
+// the cluster network.
+// Note: this CRD is managed using a singleton cluster-wide CR, which is
+// called: `cluster`. This is enforced by the CRD's unique name in
+// combination with the fact that the CRD is cluster-scoped.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=cloudprivateipconfig,scope=Cluster
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// spec is the definition of the desired private IP request(s).
+	// +kubebuilder:validation:Optional
+	// +optional
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec,casttype=CloudPrivateIPConfigSpec"`
+	// status is the observed status of the desired private IP request(s). Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status,casttype=CloudPrivateIPConfigStatus"`
+}
+
+// CloudPrivateIPConfigSpec specifies the requested list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpec struct {
+	// ipConfigs is a unique set of tuples (node, ip). A node can have several
+	// different IPs defined by creating a separate tuple for each.
+	// The order is not strict.
+	// +kubebuilder:validation:Required
+	// +required
+	IPConfigs []CloudPrivateIPConfigSpecItem `json:"ipConfigs" protobuf:"bytes,1,opt,name=ipConfigs"`
+}
+
+// CloudPrivateIPConfigStatus specifies the requested list of private IPs and their corresponding node assignment,
+// together with their assignment conditions.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigStatus struct {
+	// ipConfigs is the unique set of .spec.ipConfigs, where each is paired with an associated condition.
+	IPConfigs []CloudPrivateIPConfigCondition `json:"ipConfigs" protobuf:"bytes,1,opt,name=ipConfigs"`
+}
+
+// CloudPrivateIPConfigSpecItem is a tuple consisting of a node name and IP address.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpecItem struct {
+	// node is the node name, as specified by the Kubernetes field: node.metadata.name
+	// +kubebuilder:validation:Required
+	// +required
```

Comment:
I would expect the node to be optional.  Since you're re-using this in status, it seems possible that you have IP with no node yet assigned.

Reply:
I don't agree here. Allowing someone to _not_ specify a node contradicts what this feature does. Where would the IP address be assigned?

Reply:
> Since you're re-using this in status, it seems possible that you have IP with no node yet assigned.

The status just copies what is defined in the spec and attaches a status indicating if that item has been successfully processed or not. The status shouldn't unset the node field, at least I never intended it to be done like that

Reply:
Your status can (and should) have an entry for an IP that has not yet been successfully applied to a node.  You need this in order to communicate your "Cloud API is processing the request".  During that timeframe, the node is not the actual node.

Also, consider the case of, when an IP moves nodes.  if the "Cloud API is processing the request", then the current status is not the status of the new node, it's actually the status of the old one.  After the IP has been removed from a given node, there may be an intermediate state where it has not been assigned to a new node yet.  Also empty.